p2p/transport/webrtc/stream.go

```diff
@@ -0,0 +1,217 @@
+package libp2pwebrtc
+
+import (
+	"net"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-libp2p/p2p/transport/webrtc/pb"
+	"github.com/libp2p/go-msgio/pbio"
+
+	"github.com/pion/datachannel"
+	"github.com/pion/webrtc/v3"
+)
+
+const (
+	// maxMessageSize is the maximum message size of the Protobuf message we send / receive.
+	maxMessageSize = 16384
+	// Pion SCTP association has an internal receive buffer of 1MB (roughly, 1MB per connection).
+	// We can change this value in the SettingEngine before creating the peerconnection.
+	// https://github.com/pion/webrtc/blob/v3.1.49/sctptransport.go#L341
+	maxBufferedAmount = 2 * maxMessageSize
+	// bufferedAmountLowThreshold and maxBufferedAmount are bound
+	// to a stream but congestion control is done on the whole
+	// SCTP association. This means that a single stream can monopolize
+	// the complete congestion control window (cwnd) if it does not
+	// read stream data and it's remote continues to send. We can
+	// add messages to the send buffer once there is space for 1 full
+	// sized message.
+	bufferedAmountLowThreshold = maxBufferedAmount / 2
+
+	// Proto overhead assumption is 5 bytes
+	protoOverhead = 5
+	// Varint overhead is assumed to be 2 bytes. This is safe since
+	// 1. This is only used and when writing message, and
+	// 2. We only send messages in chunks of `maxMessageSize - varintOverhead`
+	// which includes the data and the protobuf header. Since `maxMessageSize`
+	// is less than or equal to 2 ^ 14, the varint will not be more than
+	// 2 bytes in length.
+	varintOverhead = 2
+)
+
+type receiveState uint8
+
+const (
+	receiveStateReceiving receiveState = iota
+	receiveStateDataRead               // received and read the FIN
```

Comment:
"DataRead" doesn't sound like a terminal state to me. It sounds like we read some data, but we could read more. If this is the final state and we expect no more reads, we should call it something else. Maybe "receiveStateDone"

Reply:
Same comment for sendStateDataSent

Reply:
This terminology is taken straight from RFC 9000: https://datatracker.ietf.org/doc/html/rfc9000#section-3.